distribution/xfer/download.go

```diff
@@ -70,7 +86,7 @@ type DownloadDescriptor interface {
 	// before).
 	DiffID() (layer.DiffID, error)
 	// Download is called to perform the download.
-	Download(ctx context.Context, progressOutput progress.Output) (io.ReadCloser, int64, error)
+	Download(ctx context.Context, progressOutput progress.Output, cacheDir string) (io.ReadCloser, int64, error)
```

Comment:
hmm... the descriptor is created in the v2Puller before the download starts, but the v2Puller doesn't own any of the concurrency control, so it doesn't have enough information to generate the cachedir paths and get/release them correctly based on if image pulls as a whole succeed or not. It creates a descriptor for all layers regardless of whether they are going to be pulled or not. We need to get/release the cachedir during the download process, so this gets weird... If we wanted to resolve the potential cache dir during descriptor creation time, we would need to construct the chain and use the ChainID for the cache dir... the whole thing becomes more complex IMO. How much do you care about this? Changing this might require changing a lot of this PR.